tools/haskell_module_worker/src/Main.hs

```diff
@@ -0,0 +1,226 @@
+  where
+    -- If there is a communication error with the worker, we let bazel know.
+    reportFailure w wr e = do
+      writeChan outChan $
+        writeWorkResponse pc (wrRequestId wr) 1 $
+          "Communication with background worker " ++ show (wId w) ++ " failed: " ++ show (e :: SomeException)
+      throwIO e
+
+    getWorker k =
+      modifyMVar lruV $ \lru -> do
+        (lru', w) <- LRUCache.get k (readIORef . wIdle) lru
+        writeIORef (wIdle w) False
+        return (lru', w)
+
+    -- If there is an exception when interacting with the worker
+    -- we shut it down and spawn a new one.
+    dropWorkerOnError w = modifyMVar_ lruV $ LRUCache.discard ((wId w ==) . wId)
+
+    statusExitCode = \case { Succeeded{} -> 0; _ -> 1 }
+
+    statusOutput = \case
+      Succeeded logs -> unlines $ intersperse "" logs
+      CompileErrors logs errs -> unlines $ intersperse "" $ logs ++ errs
+      NonHaskellInputs files ->
+        unlines $ "haskell_module_worker error: non-haskell inputs:" : files
+      NonOneShotCompilation ->
+        "haskell_module_worker error: only on-shot compilation (-c) is supported"
+
+-- The key of the request is computed from the compiler arguments
+-- in the request
+requestKey :: Maybe FilePath -> WorkRequest -> IO Fingerprint
+requestKey msandboxDir0 wr = do
+    dbs <- readPackageEnv msandboxDir0 (wrArgs wr)
+    let dbs2 = collectPackageDbs (wrArgs wr)
+        odir = collectODir (wrArgs wr)
+    return $ fingerprintFingerprints $ map fingerprintString (odir ++ dbs ++ dbs2)
+  where
+    readPackageEnv :: Maybe FilePath -> [String] -> IO [String]
+    readPackageEnv msandboxDir args =
+        case drop 1 $ dropWhile ("-package-env" /=) args of
+          envFile : _ -> do
+            s <- readFile $ maybe envFile (</> envFile) msandboxDir
+            return $ takeDirectory envFile : [ x | x <- lines s, "package-db" `isPrefixOf` x ]
+          [] ->
+            return []
+
+    collectODir :: [String] -> [String]
+    collectODir ("-odir" : x : _) = [x]
+    collectODir (_ : xs) = collectODir xs
+    collectODir [] = []
+
+    collectPackageDbs :: [String] -> [String]
+    collectPackageDbs ("-package-db" : x : xs) = x : collectPackageDbs xs
+    collectPackageDbs (_ : xs) = collectPackageDbs xs
+    collectPackageDbs [] = []
+
+-- | This is the main entry point for background workers
+workerLoop :: IO ()
+workerLoop = do
+    inputBS0 <- ByteString.Lazy.getContents
+    cwd <- getCurrentDirectory
+    runSession $ go cwd inputBS0
+  where
+    go cwd bs = do
+      case Binary.decodeOrFail bs of
+        Right (rest, _, wr) -> do
+          liftIO $ forM_ (wrSandboxDir wr) (setCurrentDirectory . (cwd </>))
+          st <- compile (wrArgs wr) (wrVerbosity wr)
+          liftIO $ do
+            ByteString.Lazy.putStr (Binary.encode st)
+            hFlush stdout
+          go cwd rest
+        Left (_, _, err) ->
+          liftIO $ hPutStrLn stderr $ "Error decoding work request: " ++ err
```

Comment:
The background worker terminates since it doesn't call recursively to `go`, this closes stdout, which lets the main worker know that something went wrong.